driver-core/src/main/com/mongodb/internal/operation/ClientBulkWriteOperation.java

```diff
@@ -0,0 +1,1051 @@
+            }
+            writer.writeInt64("txnNumber", localEncodedBatchInfo.txnNumber);
+        }
+
+        void encodeWriteModel(
+                final BsonWriter writer,
+                final ClientWriteModel model,
+                final int modelIndexInBatch,
+                final int namespaceIndexInBatch) {
+            assertNotNull(encodedBatchInfo).modelsCount++;
+            writer.writeStartDocument();
+            if (model instanceof ConcreteClientInsertOneModel) {
+                writer.writeInt32("insert", namespaceIndexInBatch);
+                encodeWriteModelInternals(writer, (ConcreteClientInsertOneModel) model, modelIndexInBatch);
+            } else if (model instanceof ConcreteClientUpdateManyModel) {
+                writer.writeInt32("update", namespaceIndexInBatch);
+                encodeWriteModelInternals(writer, (ConcreteClientUpdateManyModel) model);
+            } else if (model instanceof ConcreteClientUpdateOneModel) {
+                writer.writeInt32("update", namespaceIndexInBatch);
+                encodeWriteModelInternals(writer, (ConcreteClientUpdateOneModel) model);
+            } else if (model instanceof ConcreteClientReplaceOneModel) {
+                writer.writeInt32("update", namespaceIndexInBatch);
+                encodeWriteModelInternals(writer, (ConcreteClientReplaceOneModel) model);
+            } else if (model instanceof ConcreteClientDeleteManyModel) {
+                writer.writeInt32("delete", namespaceIndexInBatch);
+                encodeWriteModelInternals(writer, (ConcreteClientDeleteManyModel) model);
+            } else if (model instanceof ConcreteClientDeleteOneModel) {
+                writer.writeInt32("delete", namespaceIndexInBatch);
+                encodeWriteModelInternals(writer, (ConcreteClientDeleteOneModel) model);
+            } else {
+                throw fail(model.getClass().toString());
+            }
+            writer.writeEndDocument();
+        }
+
+        private void encodeWriteModelInternals(final BsonWriter writer, final ConcreteClientInsertOneModel model, final int modelIndexInBatch) {
+            writer.writeName("document");
+            Object document = model.getDocument();
+            @SuppressWarnings("unchecked")
+            Encoder<Object> documentEncoder = (Encoder<Object>) codecRegistry.get(document.getClass());
+            assertNotNull(encodedBatchInfo).insertModelDocumentIds.compute(modelIndexInBatch, (k, knownModelDocumentId) -> {
+                IdHoldingBsonWriter documentIdHoldingBsonWriter = new IdHoldingBsonWriter(
+                        writer,
+                        // Reuse `knownModelDocumentId` if it may have been generated by `IdHoldingBsonWriter` in a previous attempt.
+                        // If its type is not `BsonObjectId`, we know it could not have been generated.
+                        knownModelDocumentId instanceof BsonObjectId ? knownModelDocumentId.asObjectId() : null);
```

Comment:
Running unified tests with coverage enabled, there are no occurrences of  ` knownModelDocumentId instanceof BsonObjectId` evaluating to `true`.   Is there a prose test for this, or if not should we create one?  Or could a unified test be written for it?

Reply:
Done in 1c3c590ce10bb5f4682ddaf2e3596636ca023700.